### ISSUE TITLE
feat: Produce Declarative Shadow DOM with Hydrate

### ISSUE THREAD

**Issue by mayerraphael** [ionic-team/stencil#4010]
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/stencil/blob/main/.github/CONTRIBUTING.md).
- [X] I agree to follow the [Code of Conduct](https://github.com/ionic-team/stencil/blob/main/CODE_OF_CONDUCT.md).
- [X] I have searched for [existing issues](https://github.com/ionic-team/stencil/issues) that already include this feature request, without success.

### Describe the Feature Request

Add a feature to the hydrate package so the `renderToString`  function produces static html using Declarative Shadow DOMs' <template shadowroot="open"> tag instead of the current custom solution.

Terminology used:
Declarative Shadow DOM = DSD.

Resources:
https://caniuse.com/declarative-shadow-dom
https://github.com/mfreed7/declarative-shadow-dom

### Describe the Use Case

The current hydrate pacakge is good, thanks for that as we all know how hard WebComponents and SSR are, but with the implementation of DSD in Webkit and easy Polyfils, the hydrate script should have an option which produces Declarative Shadow DOM.

Please keep in mind that my internal knowledge of Stencil is limited, I take some assumptions with the benefits this feature could have.

What are the benefits:
- The browser automatically creates a shadow-root fragment from `<template shadowroot="open">`. For Firefox, a simple polyfill can be used.
- This also means that the WebComponents can be 100% compatible with React/SSR, as React only sees the same tree as the one specified in the template. No hydration warnings anymore.
- No extra css-scoping and de-scoping required. Just place the css as is inside a <style> tag inside the `<template>`. 
- `renderToString`  may not require the full html anymore, as no mutations to `<head>` are needed. It would only need the code of the component(s).
- Hydration does need not a full re-rendering. The contents of the DSD are still present in the #shadow-root fragment.
- Currently this means that Stencil must hydrate components before any other framework does, otherwise it would cause for example React Hydration Errors as the DOM does not match.

### Describe Preferred Solution
- "Stencil Hydrate" should allow to render components using Declarative Shadow DOM
- "Stencil Hydrate"s' hydration logic should work with existing ShadowRoot and Nodes upon initialization to avoid unecessary re-rendering.
- Styles should be usable without the current custom scoping solution.

### Additional Information

I dont have any information on how Firefox plans to support DSD, but for cross-framework-compatible SSR i see no way around DSD.

Edit: Firefox DSD implementation is finished: https://bugzilla.mozilla.org/show_bug.cgi?id=1712140

**Comment by rwaskiewicz** [comment]
Thanks for the detailed request @mayerraphael! Feature requests like these are very much appreciated. This is something we've been keeping an eye on, and don't have an existing issue/feature request for. I'm going to label this to try to gauge interest in something like this, which helps from a prioritization standpoint. Thanks again!

**Comment by mayerraphael** [comment]
@rwaskiewicz Any update on when or if this is planned? With currently 16 upvotes this seems to get spark some interest.
I guess this would also simplify the current custom hydration logic massivly (maybe some problems that needed to be fixed with [PR 2938](https://github.com/ionic-team/stencil/pull/2938))

**Comment by rwaskiewicz** [comment]
@mayerraphael It's moved up on our backlog to do some spike work around this, but I don't have a concrete timeline to share at this time

**Comment by mayerraphael** [comment]
@rwaskiewicz 

Just wanted to say that Porsche is also using Stencil, but they wrote their own SSR solution using Declarative ShadowDOM:

https://github.com/porsche-design-system/porsche-design-system/blob/aa8177ff61c916c8c6872844650fd2b0bfaf1d23/packages/components/scripts/generateDSRComponents.ts#L121

**Comment by benelan** [comment]
Not sure if this is possible, but it would be awesome if generating declarative shadow roots happened via a build flag rather than as an output of `renderToString`.

Using `renderToString` with SSR frameworks requires swapping in a custom server in a lot of cases, which removes some of the frameworks' benefits. For example, [Next.js says](https://nextjs.org/docs/pages/building-your-application/configuring/custom-server):

> Before deciding to use a custom server, please keep in mind that it should only be used when the integrated router of Next.js can't meet your app requirements. A custom server will remove important performance optimizations, like serverless functions and Automatic Static Optimization.
>
> A custom server cannot be deployed on Vercel.

Compiling the Stencil components directly to custom elements with declarative shadow roots should theoretically remove the need for middleware like `renderToString`.

**Comment by mayerraphael** [comment]
@benelan This would be the job of the React (Wrapper) component then.

I've created a repo which shows how DSD works with a handwritten WebComponent and NextJS: https://github.com/mayerraphael/nextjs-webcomponent-hydration

It works **WITHOUT** a custom `server.js`.

The DSD cannot be part of the WebComponent because if you use the custom element directly in NextJS, like `<my-component />`, it does not call a render function on that component, but renders the tag as is.

Now a React-based wrapper components will be called by NextJS and can therefor inject the DSD as seen in my example.
Porsche does the same: their custom SSR code generates React components from the Stencil source files.

I will extend my example soon using StencilJS and try to create a generic wrapper around Stencil components so they can be used with SSR.

**Comment by benelan** [comment]
Thanks for the info and sample @mayerraphael! I haven't worked with DSDs yet, but my understanding is the templates will be rendered on the server, and then the custom elements will hydrate on the client. This [Chrome article](https://developer.chrome.com/articles/declarative-shadow-dom/#component-hydration
) talks about that a bit.

> The DSD cannot be part of the WebComponent because if you use the custom element directly in NextJS, like `<my-component />`, it does not call a render function on that component, but renders the tag as is.

Do you mean it literally renders the component tag as text or only renders the template?

If it's the latter, that may be a React-specific issue with their VDOM implementation not re-rendering once the custom element hydrates. [Stencil provides a React wrapper output target](https://stenciljs.com/docs/react#adding-the-react-output-target) that should be able to help wire up up the DSDs and force a rerender after hydration, if that's the issue.

I just used NextJS as an example but there are plenty of other SSR/SSG frameworks like Astro/SveleKit/NuxtJS that won't run into React's web component weirdness. 

Note: I was able to prerender Stencil components in NextJS using `getStaticProps` instead of a custom server at one point, but it was pretty fragile. [Here is the sample](https://github.com/benelan/calcite-samples/tree/benelan/get-static-props/ssr-frameworks/next) if anyone is curious.

**Comment by mayerraphael** [comment]
> Thanks for the info and sample @mayerraphael! I haven't worked with DSDs yet, but my understanding is the templates will be rendered on the server, and then the custom elements will hydrate on the client. This [Chrome article](https://developer.chrome.com/articles/declarative-shadow-dom/#component-hydration) talks about that a bit.
> 

Exactly, the goal is to render the `<template shadowrootmode>` Tag, which the Browser converts on parsing the HTML to a ShadowRoot Fragment. This means we have a working Shadow Root even before the component hydrates on the client.

> > The DSD cannot be part of the WebComponent because if you use the custom element directly in NextJS, like `<my-component />`, it does not call a render function on that component, but renders the tag as is.
> 
> Do you mean it literally renders the component tag as text or only renders the template?

I mean the first. If you have a basic HtmlElement WebComponent, you specifiy it in your code by tag, like `<my-component>`. There is no import or class related for NextJS to instantiate that class and call render() on. So it cannot now the "internals" of the component. It only places the tag as is in the HTML. It is like you write a `<div>` tag.

You can try it yourself. Just normally place your custom element by tag name (<my-component>) inside a NextJS page and disable JavaScript in the Browser or check the generated html. The components content will not be visible (only the children, as they can be rendered by NextJS depending if they are webcomponents too or not) until the component hydrates.

> If it's the latter, that may be a React-specific issue with their VDOM implementation not re-rendering once the custom element hydrates. [Stencil provides a React wrapper output target](https://stenciljs.com/docs/react#adding-the-react-output-target) that should be able to help wire up the DSDs and force a rerender after hydration, if that's the issue.
> 

We are not on the client yet, that is another topic with problems regarding [Reconciliation React Issue](https://github.com/facebook/react/issues/26071). This can be worked around by providing different VDOM on client and on server, as seen [here](https://github.com/mayerraphael/nextjs-webcomponent-hydration/blob/main/components/StencilWrapper.tsx#L93). This is basically a custom React Wrapper (like the React Stencil Output Target creates, but more incomplete) which handles the DSD on the server and no-dsd on the client so React does not create hydration errors because of VDOM mismatches.

> I just used NextJS as an example but there are plenty of other SSR/SSG frameworks like Astro/SveleKit/NuxtJS that won't run into React's web component weirdness.
> 
They also do. The problem is always the same. With a native WebComponent, you need to integrated it into the Framework so it can render the contents of the WebComponent on the server, as WebComponents are a client side only construct by nature (with the CustomElementRegistry and so on). So you need to "tell" the framework on how to run your component and what to do with the "render()" result.

Stencil uses a custom [VNode format here](https://github.com/ionic-team/stencil/blob/cb3eac3a2b7795a5c8d3f9100812337d1a8d8a9d/src/declarations/stencil-public-runtime.ts#L621), derived from Preact. Because of that we need to convert the result of the `render()` call to another VNode format, whatever you use for rendering.

This is what i'm doing [here](https://github.com/mayerraphael/nextjs-webcomponent-hydration/blob/main/components/StencilWrapper.tsx#L75). Instantiate the Component class (pass an empty hostref WeakMap), apply props and call the render method. Then i convert the StencilJS VNodes to Preact VNodes, so i can render them on the server. Would also be possible to convert them directly to React VNodes, but i found using preact-render-to-string easier.

> Note: I was able to prerender Stencil components in NextJS using `getStaticProps` instead of a custom server at one point, but it was pretty fragile. [Here is the sample](https://github.com/benelan/calcite-samples/tree/benelan/get-static-props/ssr-frameworks/next) if anyone is curious.

Yeah here you are rendering the components practically in a browser stack and generate static pages from it. Not viable for any dynamic content.

**Comment by benelan** [comment]
That all makes sense, thanks for taking the time to write up the response! I'm looking forward to playing with DSDs more.

I think my initial suggestion of generating web components with DSDs during Stencil's compilation is still valid though. It sounds like we are in agreement that Stencil's framework wrapper output targets should hold the responsibility of wiring up the DSDs. So `renderToString` shouldn't be required at all if the web components already have DSDs.

This is all theoretical, I have no idea if this is possible with Stencil's component architecture or compilation process. And if it is possible, I'm sure it would be a much larger undertaking than switching `renderToString` to generate DSDs.

However, there would be many benefits to removing the need for the component library consumers to use `renderToString`. It would reduce their build time, which can be pretty significant for very large applications. It would also reduce the steps required for users to get the component library set up, which will differ depending on the SSR framework and may not even be possible/stable in some frameworks.

**Comment by alicewriteswrongs** [comment]
@benelan I'm curious what you're proposing with generating components with DSD during Stencil compilation. My understanding is that DSD is an "in-html" construct, so at a minimum Stencil would have to start distributing some sort of artifact that's a bit different than what we distribute now for the hydrate app output target (or any of the others for that matter).

There would be some advantages to having Stencil produce some sort of DSD 'intermediate representation' itself (whatever that might look like), namely that then the framework wrappers would be able to all consume a unified format instead of having to DIY it, and additionally Stencil developers who don't use a framework would also likely be able to find a way to take advantage of DSD in their applications.

Where it gets a bit interesting is what that DSD 'IR' would look like, as DSD is an in-HTML construct. Since the rendered output of a Stencil component is generally going to be dependent on the props it receives, and prop values wouldn't be known at compile-time, we wouldn't be able to just produce a little html fragment. So maybe a per-component function called something like `renderDeclarativeShadowDOM` which takes the component's props as an argument and returns an HTML string like

```html
<my-component>
  <template shadowrootmode="open">
    <style>{{ component styles }}</style>
    {{ rendered component html }}
  </template>
</my-component>
```

I would think that a function like that could be called in the framework wrappers or directly for both SSR and SSG use-cases to cover both dynamic components which need per-page-view props and static components which only need to be rendered once per page, but the details of how exactly that would happen would be a bit of a downstream concern from Stencil itself.

That feels like something that would give a lot of flexibility for generating a DSD for a given Stencil component, while also not bringing the responsibility for _all_ of this into Stencil itself.

Is that along the lines of what you're thinking @benelan or something else?

**Comment by benelan** [comment]
@alicewriteswrongs I hadn't considered prop values when I was thinking about this, but your potential solution sounds great. 

My main goal is for our component library consumers to be able to use the framework wrappers in both SSR and SPA frameworks without needing to use renderToString on their end. It sounds like the solution you're envisioning would allow for that, which would be awesome!

**Comment by benelan** [comment]
In case you haven't seen yet, Lit has an [experimental SSR solution](https://github.com/lit/lit/tree/main/packages/labs/ssr#readme) that takes advantage of DSDs. It could be helpful to see how they did it while you plan Stencil's implementation.

**Comment by alicewriteswrongs** [comment]
Definitely something to keep considering and noodling on here! I suspect you're right that for some frameworks it would probably be relatively doable to generate one component that will work for both a server-side and client-side case if Stencil exported a first-class way to generate a DSD (i.e. a `<template>` tag) for the elements _but_ I will also add a big caveat that at present we haven't yet tried to do any of this so when any implementation actually starts things could always end up being a good deal more complicated than we think.

When I think in my head about how this could work in react + nextjs, for instance, I can kind of visualize a path through where Stencil exports some way to generate a DSD from props, a React component wrapper either calls that function _or_ static JSX is generated based on that (or something along those lines) and then Stencil's component runtime is modified to support gracefully taking over components initialized with DSD, allowing a React component wrapping a Stencil component to be server-rendered _or_ client-rendered, with the runtime gracefully "doing the right thing" in either case, but that's actually a lot of pieces to get working together so a lot of opportunities for things to be more difficult than would be ideal 😅

As a sort of meta-comment on this issue, the Stencil team is looking to do more exploration and research about this soon and while I can't promise a definite timeline it is definitely something we're looking at

**Comment by denyo** [comment]
> @rwaskiewicz
> 
> Just wanted to say that Porsche is also using Stencil, but they wrote their own SSR solution using Declarative ShadowDOM:
> 
> https://github.com/porsche-design-system/porsche-design-system/blob/aa8177ff61c916c8c6872844650fd2b0bfaf1d23/packages/components/scripts/generateDSRComponents.ts#L121

Maybe to give some more details regarding our solution:
- we do framework wrapper generation ourselves for Angular, React and Vue
- the SSR support for Next.js and Remix is an extension of the React wrappers, so it currently only works with React SSR
- it's quite messy, especially the countless replacements and transformations done in `generateDSRComponents.ts` that transform the raw render functions of the Stencil components into something usable by React
- the switch between what is rendered on server side / client side is based on `process.browser` which needs to be injected/replaced with `true | false` during build: https://github.com/porsche-design-system/porsche-design-system/blob/41a6f853db436ca8f0d286be584b96ca0ec2b4d0/packages/components/scripts/wrapper-generator/NextJsReactWrapperGenerator.ts#L69-L79
- then we rely on dead code elimination to not ship the DSR wrappers to the client
- React SSR wrappers are shipped via sub package `ssr` of https://www.npmjs.com/package/@porsche-design-system/components-react

**Comment by mayerraphael** [comment]
> * we also had to patch stencil core to make bootstrapping (you may also call it hydration) work since it does not have to create a shadow root, when there is already a DSR

@denyo We had to do the same. Calling attachShadow on an existing ShadowRoot will clear the ShadowRoot. Causing a full re-render of the components on the client.

`this.shadowRoot ?? self.attachShadow({ mode: 'open' });`

> * upon client side initialization the web component just rerenders everything within which leads to flickering when there are nested web components since it does not do real hydration which would just add event listeners and not rerender

Weird, we do not experience this. At least not with patching attachShadow. We also did patch `oldVNode` in `renderVdom` with the children of the DSD ShadowRoot. This causes Stencil to only patch the existing DOM and not do duplicate insert of nodes because it thinks they do not exist already.

We are quite happy with the current solution we have, even though its as you already said: "quite messy". But we had to get away from Stencil Hydrate because of the many bugs it has (even with v4.1+ and experimentalSlotFixes).

Creating native React components from our Stencil components and patching the Stencil runtime reduced the "Total Block Time" from 400ms (Stencil Hydrate) to 10ms (our DSD solution with patched runtime). Also CLS is reduced.

Before:
![image](https://github.com/ionic-team/stencil/assets/25869389/b851ff83-d1b6-46f5-97da-002c7b98c624)

After:
![image](https://github.com/ionic-team/stencil/assets/25869389/169108f2-7c83-44f4-8846-37dae68791f6)

This causes a difference of  ~25 Lighthouse Score points (from 65 to 90), whoever cares about that.

@rwaskiewicz @alicewriteswrongs If Stencil could do "real" hydration, taking existing ShadowRoot nodes into account, this would be a very good first step and i guess also an required step to a new DSD hydration solution. It could also work hand in hand with the current Stencil Hydrate solution.

**Comment by igorlino** [comment]
FYI support for Declarative Shadow DOM (DSD) has been merged and is already part of the HTML spec  under https://github.com/whatwg/html/pull/9538

which was mentioned in this article: https://developer.chrome.com/en/articles/declarative-shadow-dom/

It would be a dream for a clean out-of-the-box way to integrate Next.js with Stencil web components without workarounds or performance penalties.

No qualified nor out-of-the-box anwers under: 
- https://stackoverflow.com/questions/76769253/how-to-integrate-stencil-web-components-in-nextjs-implementing-ssr-not-csr
- https://stackoverflow.com/questions/76211672/using-web-components-in-react-sever-components

**Comment by christian-bromann** [comment]
We have implemented support for this in #5792. Please give it a try by installing our dev build for this:

```sh
npm install @stencil/core@4.18.3-dev.1718220868.532557e
```

And provide feedback. Thanks!

**Comment by christian-bromann** [comment]
Thanks @mayerraphael for taking a look at testing it out, I've received this feedback from my peers as well and will investigate.

**Comment by christian-bromann** [comment]
@mayerraphael can you provide a minimal reproducible example by any chance?

**Comment by christian-bromann** [comment]
> I faked the SSR part

This is not supported as Stencil relies on HTML comments created during the serialization process. Without them it won't be able to reconcile the elements properly. Applying the following patch made the example work:

```diff
diff --git a/server.js b/server.js
index c75334c..606385f 100644
--- a/server.js
+++ b/server.js
@@ -2,29 +2,17 @@ import express from 'express';
 const app = express();
 const port = 3333;

+import { renderToString } from './hydrate/index.mjs';
+
 app.use("/static", express.static("."));


-app.get('/', (req, resp) => {
-  resp.send(`
-        <my-component>
-           <template shadowrootmode="open">
-                <style>
-                    :host {
-                        display: block;
-                    }
-                </style>
-                <div>
-                    <slot></slot>
-                </div>
-           </template>
-           <span>Hello SSR.</span>
-        </my-component>
-        <script type="module">
-            import {defineCustomElements}  from "./static/loader/index.js";
-            defineCustomElements().catch(console.error);
-        </script>
-    `);
+app.get('/', async (req, resp) => {
+  const { html } = await renderToString(`<my-component>Hello SSR.</my-component>`, {
+    serializeShadowRoot: true,
+    fullDocument: false
+  });
+  resp.send(html);
 });


diff --git a/stencil.config.ts b/stencil.config.ts
index c58df8b..0e98e8e 100644
--- a/stencil.config.ts
+++ b/stencil.config.ts
@@ -19,6 +19,9 @@ export const config: Config = {
       type: 'www',
       serviceWorker: null, // disable service workers
     },
+    {
+      type: 'dist-hydrate-script',
+    }
   ],
   testing: {
     browserHeadless: "new",
```

**Comment by christian-bromann** [comment]
@mayerraphael thanks a lot for your feedback!

> The whole reason for Declarative Shadow DOM is to be able to let the HTML/CSS inside the component as-is so no reconstruction is required, just "real hydration" (attaching handlers).

This makes sense. I would love to track this in a separate issue as I believe it requires a new approach on how we hydrate components in general. We already had conversations within them team of getting rid of the HTML comments but this may be a breaking change that we have to carefully evaluate. Hence I suggest to track this as a next step.

> The current solution would require, again, to first serialize the whole React/NextJS page and then pass everything to renderToString.

How so? You can serialize/hydrate individual components by setting `fullDocument: false` as an option. In fact, my next step is to enhance the StencIl React Output target to support Next.js application using this approach.

Keep an 👁️ on #5831 , I would love to get your feedback on this. I hope to have something to test for you within the next 1/2 weeks.

**Comment by christian-bromann** [comment]
> How would serialization work in this case?

We are working on an enhancement on the React output target which would provide better support for Next.js. The approach is to wrap the Stencil component to use the `renderToString` method on server side and the react output target on client side. We have a POC that verified this approach to be working fairly well. This would indeed mean we will have logic that transforms the Stencil component and its props into a string internally.

> Your solution, extending the React output target: Does this mean inside the wrapper the contents are serialized to string and some kind of dangerouslySetInnerHTML element is rendered?

Correct, it may look something like this:

```ts
import { renderToString } from 'component-library/hydrate';

export const MyButton = typeof globalThis.window !== 'undefined'
    /**
     * export React output target
     */
    ? createComponent<MyButtonElement, MyButtonEvents>({
        tagName: 'my-button',
        elementClass: MyButtonElement,
        react: React,
        events: {
            onMyFocus: 'myFocus',
            onMyBlur: 'myBlur'
        } as MyButtonEvents,
        defineCustomElement: defineMyButton
    })
    :
    /**
     * export serialized version of component for server side rendering
     */
    async (props: React.PropsWithChildren<{}>) => {
        const { html } = await renderToString(
            `<my-button ${reactPropsToStencilAttrs(props)}></my-button>`,
            {
                serializeShadowRoot: true,
                fullDocument: false
            }
        );
        return (
            <my-button suppressHydrationErrors>
                <template shadowrootmode="open" dangerouslySetInnerHTML={{
                    __html: html
                }} />
				{ props.children }
            </my-button>
        )
    }
```

> But as s-id is incremental and unique for each rendering and the c-ids are dependent on s-id (
 part) and conditional parts inside the component, this can never be done at compile time but only at runtime?

Yes, these are required at runtime for Stencil to reconcile the VDOM correctly. They are indeed random which is why the `suppressHydrationErrors` flag has to be set. As mentioned before we are looking for better solutions, e.g. not being dependent on any custom attributes or html comments for reconciliation but we will approach this one step at a time.

Thanks for all your great feedback so far.

**Comment by tanner-reits** [comment]
The initial work for this was included in today's [v4.19.0 release](https://github.com/ionic-team/stencil/releases/tag/v4.19.0)! Please let us know of any bugs or complications in a _new_ issue.
